### PR TITLE
Testing gui

### LINKED ISSUE #22: Testing
Testing is currently very manual and lacks options. A testing GUI that allows generating demo situations may be very useful.

### PR DESCRIPTION
**Pull request by jeiang**
Closes #22. 

Will be undrafted when #21 is closed.

### CODE REVIEW

File: Physics/Assets/Scripts/IP Transmission/Sender.cs [hunk 1]
```diff
@@ -151,7 +151,41 @@ public void SendAsync(string data)
...
+                {
+                    Debug.LogError("Socket Exception occurred while trying to connect! " +
+                        $"Retrying {++retryCount}/{_maxRetries}. " +
+                        $"Error: {se.SocketErrorCode}. " +
+                        $"Error Code: {se.ErrorCode}.");
+                    if (se.ErrorCode != 10061) // if not ConnectionRefused quit
+                    {
+                        Debug.LogError("Aborting...");
+                        return;
+                    }
```
Comment (D-Pino) [inline Physics/Assets/Scripts/IP Transmission/Sender.cs]: Stray whitespace in this line

File: Physics/Assets/Scripts/IP Transmission/Sender.cs [hunk 2]
```diff
@@ -151,7 +151,41 @@ public void SendAsync(string data)
                 return;
             }
 
-            _sender.Connect(_remoteEndPoint);
+            int retryCount = 0;
+            while (retryCount < _maxRetries)
```
Comment (D-Pino) [inline Physics/Assets/Scripts/IP Transmission/Sender.cs:155]: retryCount is only incremented in the catch for SocketException. If a different type of exception is repeatedly encountered, will this loop indefinitely? Is that the intended behaviour?

File: Physics/Assets/Editor/Testing Script.cs [hunk 3]
```diff
@@ -1,116 +1,322 @@
...
+        private ExportScene.ExportType _exportType = ExportScene.ExportType.None;
+        private bool _exportTestRenders = false;
+        private bool _useExplosion = true;
+        private Vector3 _explosionOrigin = Vector3.zero;
+        private float _explosionUpwardsModifier = 10;
+        private ForceMode _forceType = ForceMode.Impulse;
+        private Vector2 _minMaxForce = new Vector2(1, 10);
+        private Vector2Int _renderResolution = new Vector2Int(1920, 1080);
+        private Vector2Int _rendererOutputResolution = new Vector2Int(1920, 1080);
+        private string _rendererOutputFolder = "";
```
Comment (D-Pino) [inline Physics/Assets/Editor/Testing Script.cs:29]: This default throws an error when a renderer output folder is not specified. Probably better to just ignore it, or log a warning/regular message instead.

**Review by D-Pino — APPROVED** [review]
LGTM, left a couple minor comments, but nothing broken, so approving anyway

File: Physics/Assets/Scripts/IP Transmission/Sender.cs (hunk 2 above)
Comment (jeiang) [reply Physics/Assets/Scripts/IP Transmission/Sender.cs:155]: Was missing a return and continue to handle the exit. Other exceptions were supposed to exit immediately. Fixed in next commit.

File: Physics/Assets/Editor/Testing Script.cs (hunk 3 above)
Comment (jeiang) [reply Physics/Assets/Editor/Testing Script.cs:29]: Added default values in init function such that no error would log. Hence there will always be an output folder.

File: Physics/Assets/Scripts/IP Transmission/Sender.cs (hunk 1 above)
Comment (jeiang) [reply Physics/Assets/Scripts/IP Transmission/Sender.cs]: Removed Space.

**Pull request merged**